driver-core/src/main/com/mongodb/operation/ChangeStreamBatchCursor.java

```diff
@@ -34,18 +34,20 @@
     private final ReadBinding binding;
     private final ChangeStreamOperation<T> changeStreamOperation;
 
-    private BsonDocument resumeToken;
     private BatchCursor<RawBsonDocument> wrapped;
+    private BsonDocument resumeToken;
 
     ChangeStreamBatchCursor(final ChangeStreamOperation<T> changeStreamOperation,
                             final BatchCursor<RawBsonDocument> wrapped,
                             final ReadBinding binding) {
         this.changeStreamOperation = changeStreamOperation;
-        this.resumeToken = changeStreamOperation.getResumeToken();
         if (changeStreamOperation.getStartAfter() != null) {
-            this.resumeToken = changeStreamOperation.getStartAfter();
+            resumeToken = changeStreamOperation.getStartAfter();
+        }
+        else if (changeStreamOperation.getResumeAfter() != null) {
+            resumeToken = changeStreamOperation.getResumeAfter();
         }
-        if (this.resumeToken == null) {
+        if (resumeToken == null) {
```

Comment:
This is what concerns me most.   @jstewart-mongo, what do you think?

Reply:
And both of you, I'm curious why no tests catch it if it's a bug.

Reply:
It looks like aggregate responses are processed in the AsyncBatchCursor. These rules will be part of the async changes.

Reply:
I will add a test for the async version of the prose tests.

Reply:
It was originally implemented as per [spec in 54bda72a](https://github.com/mongodb/specifications/blob/54bda72a8974c22960e6adb0cc32ffac8ceb5832/source/change-streams/change-streams.rst).  It looks like since then there have been numerous updates to the spec for this logic.

To bring it closer to spec, add a check so see if the operation has a `startAtOperationTime` before setting it. 

---

I personally would advocate adding an extra ticket in jira for checking: 
```
The initial aggregate response had no results.
The initial aggregate response did not include a postBatchResumeToken.
```
Its has more refactoring costs - as the conversion of at least the initial batch has to be upfront and not on demand.

Reply:
`startAtOperationTimeForResume` is set in the constructor, not `startAtOperationTime`. I am concerned that if we do not set `startAtOperationTimeForResume` if `startAtOperationTime` is set, then on a resumeable operation `startAtOperationTime` may be cleared and we would not have either a resume token or an operation time.

I am unable to call `wrapped.tryHasNext()` from here, but I can check `wrapped.getPostBatchResumeToken()`. I will add that check here and retest.

Reply:
SGTM

Reply:
SGTM :+1:

Reply:
I have added the check.

Reply:
For me, it's still not clear the difference between `async` and `sync` code parts in java. I mean that probably the logic which I mentioned in the first comment can(should?) be implemented in these lines. I don't see yet why this logic should be part of `async` changes.

Also, what PR will contain `async` changes? Will we push only `sync` changes into `master`? if so, it looks like we will have not fully implemented ticket pushed to master, which is probably not so good

Reply:
I think async can be done separately in another pr - it'll be a smaller review.  Currently FLE is sync only and has a ticket to do the async work.